pkg/odo/cli/component/test.go

```diff
@@ -0,0 +1,109 @@
+package component
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	adaptercommon "github.com/openshift/odo/pkg/devfile/adapters/common"
+	devfileParser "github.com/openshift/odo/pkg/devfile/parser"
+	"github.com/openshift/odo/pkg/devfile/parser/data/common"
+	projectCmd "github.com/openshift/odo/pkg/odo/cli/project"
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	odoutil "github.com/openshift/odo/pkg/odo/util"
+	"github.com/openshift/odo/pkg/odo/util/completion"
+	"github.com/openshift/odo/pkg/odo/util/pushtarget"
+	"github.com/spf13/cobra"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// RecommendedCommandName is the recommended command name
+const TestRecommendedCommandName = "test"
+
+// TestOptions encapsulates the options for the odo command
+type TestOptions struct {
+	commandName      string
+	componentContext string
+	namespace        string
+	devfilePath      string
+	testCommand      common.DevfileCommand
+	*genericclioptions.Context
+}
+
+var testExample = templates.Examples(`
+  # Run default test command
+  %[1]s
+
+  # Run a specific test command
+  %[1]s --test-command <command name>
+
+`)
+
+// NewTestOptions creates a new TestOptions instance
+func NewTestOptions() *TestOptions {
+	return &TestOptions{}
+}
+
+// Complete completes TestOptions after they've been created
+func (to *TestOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	to.devfilePath = filepath.Join(to.componentContext, DevfilePath)
+	to.Context = genericclioptions.NewDevfileContext(cmd)
+	if !pushtarget.IsPushTargetDocker() {
+		// The namespace was retrieved from the --project flag (or from the kube client if not set) and stored in kclient when initalizing the context
+		to.namespace = to.KClient.Namespace
+	}
+
+	return
+}
+
+// Validate validates the TestOptions based on completed values
+func (to *TestOptions) Validate() (err error) {
+	devObj, err := devfileParser.ParseAndValidate(to.devfilePath)
+	if err != nil {
+		return errors.Wrap(err, "fail to parse devfile")
+	}
+	if reflect.DeepEqual(devObj.Ctx.GetApiVersion(), "1.0.0") {
+		return fmt.Errorf("'odo test' is not supported in devfile 1.0.0")
+	}
+	to.testCommand, err = adaptercommon.GetTestCommand(devObj.Data, to.commandName)
```

Comment:
I would say get the test command in the adapter and not in the cli code, to be consistent with rest of odo since all the devfile cli is doing the other way ie just pass the command name string and get it in the adapter code

Reply:
Updated